contracts/governance/Staking/modules/StakingWithdrawModule.sol

```diff
@@ -0,0 +1,212 @@
+        if (block.timestamp < until && !allUnlocked && !isGovernance) {
+            uint96 punishedAmount = _getPunishedAmount(amount, until);
+            amount -= punishedAmount;
+
+            /// @dev punishedAmount can be 0 if block.timestamp are very close to 'until'
+            if (punishedAmount > 0) {
+                require(address(feeSharing) != address(0), "FeeSharing address wasn't set"); // S08
+                /// @dev Move punished amount to fee sharing.
+                /// @dev Approve transfer here and let feeSharing do transfer and write checkpoint.
+                SOVToken.approve(address(feeSharing), punishedAmount);
+                feeSharing.transferTokens(address(SOVToken), punishedAmount);
+            }
+        }
+
+        /// @dev transferFrom
+        bool success = SOVToken.transfer(receiver, amount);
+        require(success, "Token transfer failed"); // S09
+
+        emit StakingWithdrawn(msg.sender, amount, until, receiver, isGovernance);
+    }
+
+    // @dev withdraws tokens for lock date 2 weeks later than given lock date
+    function _withdrawNext(
+        uint256 until,
+        address receiver,
+        bool isGovernance
+    ) internal {
+        if (_isVestingContract(msg.sender)) {
+            uint256 nextLock = until.add(TWO_WEEKS);
+            if (isGovernance || block.timestamp >= nextLock) {
+                uint96 stakes = _getPriorUserStakeByDate(msg.sender, nextLock, block.number - 1);
+                if (stakes > 0) {
+                    _withdraw(stakes, nextLock, receiver, isGovernance);
+                }
+            }
+        }
+    }
+
+    /**
+     * @notice Get available and punished amount for withdrawing.
+     * @param amount The number of tokens to withdraw.
+     * @param until The date until which the tokens were staked.
+     * */
+    function getWithdrawAmounts(uint96 amount, uint256 until)
+        external
+        view
+        returns (uint96, uint96)
+    {
+        _validateWithdrawParams(amount, until);
+        uint96 punishedAmount = _getPunishedAmount(amount, until);
+        return (amount - punishedAmount, punishedAmount);
+    }
+
+    /**
+     * @notice Get punished amount for withdrawing.
+     * @param amount The number of tokens to withdraw.
+     * @param until The date until which the tokens were staked.
+     * */
+    function _getPunishedAmount(uint96 amount, uint256 until) internal view returns (uint96) {
+        uint256 date = _timestampToLockDate(block.timestamp);
+        uint96 weight = _computeWeightByDate(until, date); /// @dev (10 - 1) * WEIGHT_FACTOR
+        weight = weight * weightScaling;
+        return (amount * weight) / WEIGHT_FACTOR / 100;
+    }
+
+    /**
+     * @notice Validate withdraw parameters.
+     * @param amount The number of tokens to withdraw.
+     * @param until The date until which the tokens were staked.
+     * */
+    function _validateWithdrawParams(uint96 amount, uint256 until) internal view {
```

Comment:
pls move this getter to the `StakingStorageModule.sol` to keep in line with the modules logic